### ISSUE TITLE
Certain Youtube links not producing preview (sometimes, seemingly randomly)

### ISSUE THREAD

**Issue by gittigger** [mattermost/mattermost-server#15604]
#### Summary
Certain YouTube links are not providing a preview at all
Other links, seemingly randomly work sometimes

#### Steps to reproduce
I am using Mattermost Version: 5.26.2
Pasting the following link in to a channel
https://youtu.be/pEKBQ9oUtSU

#### Expected behavior
I expect to see a preview of the webpage, 
n.b. the link WORKS on your demo server on Mattermost Version: 5.7.0

#### Observed behavior (that appears unintentional) 
Nothing at all other than the link

#### Possible fixes
No idea. I turned on debugging for the server but nothing was presented regarding pasting this link in

**Comment by gittigger** [comment]
I just attempted to upgrade to latest and it still doesn't work there, but I also just noticed that on the mobile app on android, the previews show with no problem at all.

**Comment by amyblais** [comment]
@gittigger 
What OS and version are you on?
What are your settings set to?
 - https://docs.mattermost.com/administration/config-settings.html#allow-untrusted-internal-connections-to
 - https://docs.mattermost.com/administration/config-settings.html#enable-link-previews

**Comment by gittigger** [comment]
Thanks for the quick response @amyblais 
This is on Ubuntu 18.04.5 - The version after the upgrade is now the latest: 5.27.0

grep AllowedUntrustedInternalConnections /opt/mattermost/config/config.json 
        "AllowedUntrustedInternalConnections": "",

This is set to an empty string it seems. 

grep EnableLinkPreviews /opt/mattermost/config/config.json 
        "EnableLinkPreviews": true,

True in the second instance

**Comment by amyblais** [comment]
Is this issue only happening with youtube links?
What are your image proxy settings set to https://docs.mattermost.com/administration/config-settings.html#image-proxy?

**Comment by gittigger** [comment]
As far as I'm aware yes, Only YouTube links. This used to work on my set up (mostly anyway)
Now it seems after the recent upgrades, it affects all YouTube links I've tried to add.
There's no image proxy involved here...

grep -A3 ImageProxySettings /opt/mattermost/config/config.json
    "ImageProxySettings": {
        "Enable": false,
        "ImageProxyType": "local",
        "RemoteImageProxyURL": "",

Further information that may be relevant is this setup is using nginx as a proxy to serve up TLS connections.
I have a suspicion the problem lies in this area rather than configuration of Mattermost itself. 
However what I find strange about this issue, is that ***none*** of the browsers I've tried can see the previews (I've tried Chrome, Firefox & Brave) but the android app shows them perfectly.

**Comment by amyblais** [comment]
Maybe you can try configuring a proxy under Settings>Network & Internet> Proxy and then restart Mattermost.
You could also try with setting up an atmos/camo image proxy to see if it helps.

**Comment by gittigger** [comment]
Thanks for the suggestion. Neither of those seem to help. For a while yesterday it was working without a proxy. Yet, today it is not. So this is an _intermittent_ problem which of course are always the hardest to debug. That said, do you have any recommendations on how I can debug this properly?
Tinkering and trying things can only get you so far and I'd like to actually start to understand the issue here. 
Thanks

**Comment by lieut-data** [comment]
@gittigger, do you have a Google API key configured? Without one, I wonder if Google is rate limiting requests. You can configure the key via the system console:

<img width="927" alt="image" src="https://user-images.githubusercontent.com/1023171/94440603-40374780-0178-11eb-81af-3187d3db1446.png">

**Comment by kevoops** [comment]
@lieut-data Thanks - I didn't have one configured. I do now, thanks for the additional pointer. This sounds like it might well be in the area. It would certainly cover off the intermittent nature of the problem. I now have a Google API Key configured within Mattermost. I'm going to leave it like this for a week or so now and see if it fixes up the issue. Thanks again.

**Comment by kevoops** [comment]
hmm, I'm not seeing any requests hitting the YouTube Data API, using the key that was generated. This isn't an Enterprise enabled feature only is it? How can I see the requests being made? I have debug enabled on my logs and cannot see any requests being made to YouTube

**Comment by lieut-data** [comment]
cc @hmhealey, do you have any insight on how we use the Google Developer Key? See above for context about intermittent previews, and my suggestion of configuring a Google API key to avoid a possible rate limiting.

**Comment by hmhealey** [comment]
The Google Developer Key is only needed for titling YouTube previews on mobile last I checked. It used to be used to get the title of the YouTube video on the web app as well until we switched it to rely on OpenGraph metadata.

The only setting that should matter here is that EnableLinkPreviews is true, which presumably, it is. I tried it on my local instance, and the video is showing correctly. I'm also not aware of any changes in this area specific to 5.26.
![Screen Shot 2020-09-28 at 5 27 36 PM](https://user-images.githubusercontent.com/3277310/94488090-fff3bb80-01af-11eb-8327-918afcf44d68.png)


@gittigger Did you upgrade directly from 5.7 to 5.26?

**Comment by gittigger** [comment]
The upgrade path has been 5.21 -> 5.26.2 -> 5.27
It is interesting you mention mobile @hmhealey because it is on android mobile app that these previews seem to show perfectly every time. 
This issue only seems to affect desktop browsers...

**Comment by hmhealey** [comment]
Does your server happen to be behind a firewall that prevents it from contacting the internet? The server has to be able to talk to YouTube (or any other sites you want previews for) for the preview to be displayed.

**Comment by gittigger** [comment]
No, the server has and always has had; a direct connection to the internet. Can you explain how this new server change works @hmhealey ?

**Comment by hmhealey** [comment]
The changes were actually made back in 5.12, but I could have sworn they were made more recently. Regardless, the process for showing these previews is that the server hits the URL of any posted link to see what type of content it is, and then it tells the client if its an image or a previewable webpage with whatever information it was able to gather. YouTube videos are a special case of the regular webpage previews now.

As long as the Enable Link Previews setting is enabled on the server and its outgoing connections aren't being blocked by anything, I'm not sure what would be causing previews not to show up, especially if it's only intermittently.

**Comment by gittigger** [comment]
Thanks. I can only think that this is environmental then. I'm working through a list of possible culprits I can think of and will report back when I have a better handle on this...

**Comment by gittigger** [comment]
Okay, so after going down many, many, many, dead-ends I finally am getting to the bottom of this.

Google IS temporarily blocking the access to YouTube from my server (on a Linode)
`curl http://www.youtube.com/watch?v=qj_NluriLs0
<HTML><HEAD><meta http-equiv="content-type" content="text/html;charset=utf-8">
<TITLE>302 Moved</TITLE></HEAD><BODY>
<H1>302 Moved</H1>
The document has moved
<A HREF="http://www.google.com/sorry/index?continue=http://www.youtube.com/watch%3Fv%3Dqj_NluriLs0&amp;q=EhAqAX4AAAAAAPA8kv_-SGDtGIWs6PsFIhkA8aeDS8UyO3XD8jJuLk5cBimqx3tv8vHpMgFy">here</A>.
</BODY></HTML>`

I just need to understand why...

**Comment by hmhealey** [comment]
What if you try `curl https://www.youtube.com/watch?v=qj_NluriLs0`? Perhaps the redirect is just happening because of the http vs https? I don't think that should matter for link previews though since we've tested having them run through link shorteners and such in the past.

**Comment by gittigger** [comment]
Thanks, but no it's not that. If you look at where the document has moved to you will see the problem. It was happening on secure connections too. That was just one example

**Comment by gittigger** [comment]
I _think_ I know what it is...
The DNS resolver uses systemd which does a lookup on the localhost which is configured on Linodes to be multiple servers for both IPv4 and IPv6. When I checked yesterday it seemed to be resolving youtube.com and youtu.be links to IPv6 addresses. So, it seems that Google is temporarily blacklisting servers hitting YouTube URLs over IPv6. I have now removed the /etc/resolv.conf symlink to systemd and have manually configured the DNS server to be one of the google IPv4 DNS servers. I have not seen the problem since...
